packages/aws-cdk/lib/api/cxapp/cloud-executable.ts

```diff
@@ -89,54 +90,75 @@ export class CloudExecutable {
-            }
-          }
-        }
+      if (trackVersions) {
+        // @deprecated(v2): this should honestly not be done here. The framework
+        // should (and will, shortly) synthesize this information directly into
+        // the template. However, in order to support old framework versions
+        // that don't synthesize this info yet, we can only remove this code
+        // once we break backwards compatibility.
+        await this.addMetadataResource(assembly);
       }
 
       return new CloudAssembly(assembly);
+    }
+  }
+
+  /**
+   * Modify the templates in the assembly in-place to add metadata resource declarations
+   */
+  private async addMetadataResource(rootAssembly: cxapi.CloudAssembly) {
+    if (!rootAssembly.runtime) { return; }
+
+    const modules = formatModules(rootAssembly.runtime);
+    await processAssembly(rootAssembly);
 
-      function formatModules(runtime: cxapi.RuntimeInfo): string {
-        const modules = new Array<string>();
+    async function processAssembly(assembly: cxapi.CloudAssembly) {
+      for (const stack of assembly.stacks) {
+        await processStack(stack);
+      }
+      for (const nested of assembly.nestedAssemblies) {
+        await processAssembly(nested.nestedAssembly);
+      }
+    }
 
-        // inject toolkit version to list of modules
-        // eslint-disable-next-line @typescript-eslint/no-require-imports
-        const toolkitVersion = require('../../../package.json').version;
-        modules.push(`aws-cdk=${toolkitVersion}`);
+    async function processStack(stack: cxapi.CloudFormationStackArtifact) {
+      const resourcePresent = stack.environment.region === cxapi.UNKNOWN_REGION
+        || RegionInfo.get(stack.environment.region).cdkMetadataResourceAvailable;
+      if (!resourcePresent) { return; }
 
-        for (const key of Object.keys(runtime.libraries).sort()) {
-          modules.push(`${key}=${runtime.libraries[key]}`);
+      if (!stack.template.Resources) {
+        stack.template.Resources = {};
+      }
+      if (stack.template.Resources.CDKMetadata) {
+        warning(`The stack ${stack.id} already includes a CDKMetadata resource`);
+        return;
+      }
+
+      stack.template.Resources.CDKMetadata = {
+        Type: 'AWS::CDK::Metadata',
+        Properties: {
+          Modules: modules,
+        },
+      };
+
+      if (stack.environment.region === cxapi.UNKNOWN_REGION) {
+        stack.template.Conditions = stack.template.Conditions || {};
+        const condName = 'CDKMetadataAvailable';
+        if (!stack.template.Conditions[condName]) {
+          stack.template.Conditions[condName] = _makeCdkMetadataAvailableCondition();
+          stack.template.Resources.CDKMetadata.Condition = condName;
+        } else {
+          warning(`The stack ${stack.id} already includes a ${condName} condition`);
         }
-        return modules.join(',');
       }
```

Comment:
Has any logic changed here at all or is this all refactor? I hate non-trivial refactors in bug fix PRs.
I can't say if I should or not should not scrutinize this code.

Reply:
I know, sorry. I was cleaning this up and then detected the bug. It shouldn't be fairly easy to parse, if you know how to chunk it.

The part that injects the resource is the same (`processStack` is an *extract method* refactor), the path leading up to it (`processAssembly`) is different: we now recurse through assemblies.